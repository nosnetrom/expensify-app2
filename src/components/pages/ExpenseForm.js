import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

const amountProcessor = (fullAmt) => {
  let extension = '';
  const amt = parseInt(fullAmt) / 100;
  if (fullAmt % 100 == 0) {
    extension = '.00';
  } else if (fullAmt % 10 == 0) {
    extension = '0';
  }
  return `${amt}${extension}`;
};

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      description: props.expense ? props.expense.description : '',
      note: props.expense ? props.expense.note : '',
      amount: props.expense ? amountProcessor(props.expense.amount) : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      error: '',
    };
  }
  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };
  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };
  onAmountChange = (e) => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  };
  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  };
  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }));
  };
  onSubmit = (e) => {
    e.preventDefault(); // To prevent page refresh
    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({
        error: 'Please complete the form with valid values.',
      }));
    } else {
      this.setState(() => ({ error: '' }));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note,
      });
    }
  };
  render() {
    return (
      <div className="expenseFormWrapper">
        {this.state.error && <p className="pError">{this.state.error}</p>}
        <form onSubmit={this.onSubmit}>
          <section>
            <label htmlFor="txtDescription">Description</label>
            <input
              type="text"
              name="txtDescription"
              autoFocus
              value={this.state.description}
              onChange={this.onDescriptionChange}
            ></input>
          </section>
          <section>
            <label htmlFor="txtAmount">Amount</label>
            <input
              type="text"
              name="txtAmount"
              value={this.state.amount}
              onChange={this.onAmountChange}
            ></input>
          </section>
          <section>
            <label htmlFor="date">Date</label>
            <SingleDatePicker
              date={this.state.createdAt}
              onDateChange={this.onDateChange}
              focused={this.state.calendarFocused}
              onFocusChange={this.onFocusChange}
              numberOfMonths={1}
              anchorDirection="right"
              isOutsideRange={() => false}
            />
          </section>
          <section>
            <label htmlFor="txtNote">Note</label>
            <input
              type="text"
              name="txtNote"
              value={this.state.note}
              onChange={this.onNoteChange}
            ></input>
          </section>
          <button type="submit" id="btnSubmit">
            Submit this expense
          </button>
        </form>
      </div>
    );
  }
}
