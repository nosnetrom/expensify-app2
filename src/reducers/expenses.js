const expensesReducerDefaultState = [];

// Expenses reducer
export default (state = expensesReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_EXPENSE':
      //console.log(action.expense);
      return [...state, action.expense];
    case 'EDIT_EXPENSE':
      return state.map((expense) => {
        if (expense.id === action.id) {
          // spread the existing expense object, then overwrite with the object as passed in
          return {
            ...expense,
            ...action.updates,
          };
        } else {
          return expense;
        }
      });
    case 'REMOVE_EXPENSE':
      return state.filter(({ id }) => id !== action.id);
    default:
      return state;
  }
};
