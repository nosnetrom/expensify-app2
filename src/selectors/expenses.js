import moment from 'moment';

// Redux filtering
export default (expenses, { text, sortBy, startDate, endDate }) => {
  return expenses
    .filter((expense) => {
      const createdAtMoment = moment(expense.createdAt);
      const startDateMatch = startDate
        ? startDate.isSameOrBefore(createdAtMoment, 'day')
        : true;
      const endDateMatch = endDate
        ? endDate.isSameOrAfter(createdAtMoment, 'day')
        : true;
      // const textMatch =
      //   expense.description.toLowerCase().indexOf(text.toLowerCase()) > -1;
      const textMatch = expense.description
        .toLowerCase()
        .includes(text.toLowerCase());
      // If all 3 consts == true, then we have a listable expense

      return startDateMatch && endDateMatch && textMatch;
    })
    .sort((a, b) => {
      // Use ternary function for sorting items by order
      if (sortBy === 'date') {
        return a.createdAt < b.createdAt ? 1 : -1;
      } else if (sortBy === 'amount') {
        return a.amount < b.amount ? 1 : -1;
      }
    });
};
