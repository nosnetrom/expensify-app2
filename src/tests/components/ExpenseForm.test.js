import React from 'react';
import { shallow } from 'enzyme';
import moment from 'moment';
import expenses from '../fixtures/expenses';
import ExpenseForm from '../../components/pages/ExpenseForm';

test('should render default ExpenseForm', () => {
  const wrapper = shallow(<ExpenseForm />);
  expect(wrapper).toMatchSnapshot();
});

test('should render ExpenseForm with example data', () => {
  const wrapper = shallow(<ExpenseForm expense={{ ...expenses[0] }} />);
  expect(wrapper).toMatchSnapshot();
});

test('should render ExpenseForm with error for invalid submission', () => {
  const wrapper = shallow(<ExpenseForm />);
  expect(wrapper).toMatchSnapshot(); // check the presubmitted form
  wrapper.find('form').simulate('submit', {
    preventDefault: () => {},
  });
  expect(wrapper.state('error').length).toBeGreaterThan(0);
  expect(wrapper).toMatchSnapshot();
});

test('should set expense description on change', () => {
  const value = 'My new description';
  const wrapper = shallow(<ExpenseForm expense={{ ...expenses[0] }} />);
  wrapper.find('input').at(0).simulate('change', { target: { value } });
  expect(wrapper.state('description')).toBe(value);
});

test('should set note on textarea change', () => {
  const value = 'My new note';
  const wrapper = shallow(<ExpenseForm />);
  wrapper.find('input').at(2).simulate('change', { target: { value } });
  expect(wrapper.state('note')).toBe(value);
});

test('should set amount to a new valid value', () => {
  const value = '5.99';
  const wrapper = shallow(<ExpenseForm />);
  wrapper.find('input').at(1).simulate('change', { target: { value } });
  expect(wrapper.state('amount')).toBe(value);
});

test('should not set amount when invalid value', () => {
  const value = '5.995';
  const wrapper = shallow(<ExpenseForm />);
  wrapper.find('input').at(1).simulate('change', { target: { value } });
  expect(wrapper.state('amount')).toBe('');
});

test('should call onSubmit prop for valid form submission', () => {
  const onSubmitSpy = jest.fn();
  const wrapper = shallow(
    <ExpenseForm expense={expenses[0]} onSubmit={onSubmitSpy} />
  );
  wrapper.find('form').simulate('submit', {
    preventDefault: () => {},
  });
  expect(wrapper.state('error')).toBe('');
  expect(onSubmitSpy).toHaveBeenLastCalledWith({
    description: expenses[0].description,
    amount: expenses[0].amount,
    note: expenses[0].note,
    createdAt: expenses[0].createdAt,
  });
});

test('should set new date onDateChange', () => {
  const now = moment();
  const wrapper = shallow(<ExpenseForm />);
  wrapper.find('SingleDatePicker').prop('onDateChange')(now);
  expect(wrapper.state('createdAt')).toEqual(now);
});

test('should set calendar focus on date change', () => {
  const focused = true;
  const wrapper = shallow(<ExpenseForm />);
  wrapper.find('SingleDatePicker').prop('onFocusChange')({ focused });
  expect(wrapper.state('calendarFocused')).toBe(focused);
});
