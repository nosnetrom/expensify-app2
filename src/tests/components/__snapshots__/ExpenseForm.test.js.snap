// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseForm with error for invalid submission 1`] = `
<div
  className="expenseFormWrapper"
>
  <form
    onSubmit={[Function]}
  >
    <section>
      <label
        htmlFor="txtDescription"
      >
        Description
      </label>
      <input
        autoFocus={true}
        name="txtDescription"
        onChange={[Function]}
        type="text"
        value=""
      />
    </section>
    <section>
      <label
        htmlFor="txtAmount"
      >
        Amount
      </label>
      <input
        name="txtAmount"
        onChange={[Function]}
        type="text"
        value=""
      />
    </section>
    <section>
      <label
        htmlFor="date"
      >
        Date
      </label>
      <SingleDatePicker
        anchorDirection="right"
        customCloseIcon={null}
        customInputIcon={null}
        date={"1970-01-01T00:00:00.000Z"}
        daySize={39}
        disabled={false}
        displayFormat={[Function]}
        enableOutsideDays={false}
        firstDayOfWeek={null}
        focused={false}
        hideKeyboardShortcutsPanel={false}
        horizontalMargin={0}
        id="date"
        initialVisibleMonth={null}
        inputIconPosition="before"
        isDayBlocked={[Function]}
        isDayHighlighted={[Function]}
        isOutsideRange={[Function]}
        isRTL={false}
        keepOpenOnDateSelect={false}
        monthFormat="MMMM YYYY"
        navNext={null}
        navPrev={null}
        numberOfMonths={1}
        onClose={[Function]}
        onDateChange={[Function]}
        onFocusChange={[Function]}
        onNextMonthClick={[Function]}
        onPrevMonthClick={[Function]}
        openDirection="down"
        orientation="horizontal"
        phrases={
          Object {
            "calendarLabel": "Calendar",
            "chooseAvailableDate": [Function],
            "clearDate": "Clear Date",
            "closeDatePicker": "Close",
            "dateIsUnavailable": [Function],
            "enterKey": "Enter key",
            "escape": "Escape key",
            "hideKeyboardShortcutsPanel": "Close the shortcuts panel",
            "homeEnd": "Home and end keys",
            "jumpToNextMonth": "Move forward to switch to the next month",
            "jumpToPrevMonth": "Move backward to switch to the previous month",
            "keyboardNavigationInstructions": "Press the down arrow key to interact with the calendar and
            select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
            "keyboardShortcuts": "Keyboard Shortcuts",
            "leftArrowRightArrow": "Right and left arrow keys",
            "moveFocusByOneDay": "Move backward (left) and forward (right) by one day",
            "moveFocusByOneMonth": "Switch months",
            "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week",
            "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week",
            "openThisPanel": "Open this panel",
            "pageUpPageDown": "page up and page down keys",
            "questionMark": "Question mark",
            "returnFocusToInput": "Return to the date input field",
            "selectFocusedDate": "Select the date in focus",
            "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel",
            "upArrowDownArrow": "up and down arrow keys",
          }
        }
        placeholder="Date"
        readOnly={false}
        renderCalendarInfo={null}
        renderDay={null}
        renderMonth={null}
        reopenPickerOnClearDate={false}
        required={false}
        screenReaderInputMessage=""
        showClearDate={false}
        showDefaultInputIcon={false}
        weekDayFormat="dd"
        withFullScreenPortal={false}
        withPortal={false}
      />
    </section>
    <section>
      <label
        htmlFor="txtNote"
      >
        Note
      </label>
      <input
        name="txtNote"
        onChange={[Function]}
        type="text"
        value=""
      />
    </section>
    <button
      id="btnSubmit"
      type="submit"
    >
      Submit this expense
    </button>
  </form>
</div>
`;

exports[`should render ExpenseForm with error for invalid submission 2`] = `
<div
  className="expenseFormWrapper"
>
  <p
    className="pError"
  >
    Please complete the form with valid values.
  </p>
  <form
    onSubmit={[Function]}
  >
    <section>
      <label
        htmlFor="txtDescription"
      >
        Description
      </label>
      <input
        autoFocus={true}
        name="txtDescription"
        onChange={[Function]}
        type="text"
        value=""
      />
    </section>
    <section>
      <label
        htmlFor="txtAmount"
      >
        Amount
      </label>
      <input
        name="txtAmount"
        onChange={[Function]}
        type="text"
        value=""
      />
    </section>
    <section>
      <label
        htmlFor="date"
      >
        Date
      </label>
      <SingleDatePicker
        anchorDirection="right"
        customCloseIcon={null}
        customInputIcon={null}
        date={"1970-01-01T00:00:00.000Z"}
        daySize={39}
        disabled={false}
        displayFormat={[Function]}
        enableOutsideDays={false}
        firstDayOfWeek={null}
        focused={false}
        hideKeyboardShortcutsPanel={false}
        horizontalMargin={0}
        id="date"
        initialVisibleMonth={null}
        inputIconPosition="before"
        isDayBlocked={[Function]}
        isDayHighlighted={[Function]}
        isOutsideRange={[Function]}
        isRTL={false}
        keepOpenOnDateSelect={false}
        monthFormat="MMMM YYYY"
        navNext={null}
        navPrev={null}
        numberOfMonths={1}
        onClose={[Function]}
        onDateChange={[Function]}
        onFocusChange={[Function]}
        onNextMonthClick={[Function]}
        onPrevMonthClick={[Function]}
        openDirection="down"
        orientation="horizontal"
        phrases={
          Object {
            "calendarLabel": "Calendar",
            "chooseAvailableDate": [Function],
            "clearDate": "Clear Date",
            "closeDatePicker": "Close",
            "dateIsUnavailable": [Function],
            "enterKey": "Enter key",
            "escape": "Escape key",
            "hideKeyboardShortcutsPanel": "Close the shortcuts panel",
            "homeEnd": "Home and end keys",
            "jumpToNextMonth": "Move forward to switch to the next month",
            "jumpToPrevMonth": "Move backward to switch to the previous month",
            "keyboardNavigationInstructions": "Press the down arrow key to interact with the calendar and
            select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
            "keyboardShortcuts": "Keyboard Shortcuts",
            "leftArrowRightArrow": "Right and left arrow keys",
            "moveFocusByOneDay": "Move backward (left) and forward (right) by one day",
            "moveFocusByOneMonth": "Switch months",
            "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week",
            "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week",
            "openThisPanel": "Open this panel",
            "pageUpPageDown": "page up and page down keys",
            "questionMark": "Question mark",
            "returnFocusToInput": "Return to the date input field",
            "selectFocusedDate": "Select the date in focus",
            "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel",
            "upArrowDownArrow": "up and down arrow keys",
          }
        }
        placeholder="Date"
        readOnly={false}
        renderCalendarInfo={null}
        renderDay={null}
        renderMonth={null}
        reopenPickerOnClearDate={false}
        required={false}
        screenReaderInputMessage=""
        showClearDate={false}
        showDefaultInputIcon={false}
        weekDayFormat="dd"
        withFullScreenPortal={false}
        withPortal={false}
      />
    </section>
    <section>
      <label
        htmlFor="txtNote"
      >
        Note
      </label>
      <input
        name="txtNote"
        onChange={[Function]}
        type="text"
        value=""
      />
    </section>
    <button
      id="btnSubmit"
      type="submit"
    >
      Submit this expense
    </button>
  </form>
</div>
`;

exports[`should render ExpenseForm with example data 1`] = `
<div
  className="expenseFormWrapper"
>
  <form
    onSubmit={[Function]}
  >
    <section>
      <label
        htmlFor="txtDescription"
      >
        Description
      </label>
      <input
        autoFocus={true}
        name="txtDescription"
        onChange={[Function]}
        type="text"
        value="USB stick"
      />
    </section>
    <section>
      <label
        htmlFor="txtAmount"
      >
        Amount
      </label>
      <input
        name="txtAmount"
        onChange={[Function]}
        type="text"
        value="8.99"
      />
    </section>
    <section>
      <label
        htmlFor="date"
      >
        Date
      </label>
      <SingleDatePicker
        anchorDirection="right"
        customCloseIcon={null}
        customInputIcon={null}
        date={"1970-01-12T13:46:40.000Z"}
        daySize={39}
        disabled={false}
        displayFormat={[Function]}
        enableOutsideDays={false}
        firstDayOfWeek={null}
        focused={false}
        hideKeyboardShortcutsPanel={false}
        horizontalMargin={0}
        id="date"
        initialVisibleMonth={null}
        inputIconPosition="before"
        isDayBlocked={[Function]}
        isDayHighlighted={[Function]}
        isOutsideRange={[Function]}
        isRTL={false}
        keepOpenOnDateSelect={false}
        monthFormat="MMMM YYYY"
        navNext={null}
        navPrev={null}
        numberOfMonths={1}
        onClose={[Function]}
        onDateChange={[Function]}
        onFocusChange={[Function]}
        onNextMonthClick={[Function]}
        onPrevMonthClick={[Function]}
        openDirection="down"
        orientation="horizontal"
        phrases={
          Object {
            "calendarLabel": "Calendar",
            "chooseAvailableDate": [Function],
            "clearDate": "Clear Date",
            "closeDatePicker": "Close",
            "dateIsUnavailable": [Function],
            "enterKey": "Enter key",
            "escape": "Escape key",
            "hideKeyboardShortcutsPanel": "Close the shortcuts panel",
            "homeEnd": "Home and end keys",
            "jumpToNextMonth": "Move forward to switch to the next month",
            "jumpToPrevMonth": "Move backward to switch to the previous month",
            "keyboardNavigationInstructions": "Press the down arrow key to interact with the calendar and
            select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
            "keyboardShortcuts": "Keyboard Shortcuts",
            "leftArrowRightArrow": "Right and left arrow keys",
            "moveFocusByOneDay": "Move backward (left) and forward (right) by one day",
            "moveFocusByOneMonth": "Switch months",
            "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week",
            "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week",
            "openThisPanel": "Open this panel",
            "pageUpPageDown": "page up and page down keys",
            "questionMark": "Question mark",
            "returnFocusToInput": "Return to the date input field",
            "selectFocusedDate": "Select the date in focus",
            "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel",
            "upArrowDownArrow": "up and down arrow keys",
          }
        }
        placeholder="Date"
        readOnly={false}
        renderCalendarInfo={null}
        renderDay={null}
        renderMonth={null}
        reopenPickerOnClearDate={false}
        required={false}
        screenReaderInputMessage=""
        showClearDate={false}
        showDefaultInputIcon={false}
        weekDayFormat="dd"
        withFullScreenPortal={false}
        withPortal={false}
      />
    </section>
    <section>
      <label
        htmlFor="txtNote"
      >
        Note
      </label>
      <input
        name="txtNote"
        onChange={[Function]}
        type="text"
        value="16GB"
      />
    </section>
    <button
      id="btnSubmit"
      type="submit"
    >
      Submit this expense
    </button>
  </form>
</div>
`;

exports[`should render default ExpenseForm 1`] = `
<div
  className="expenseFormWrapper"
>
  <form
    onSubmit={[Function]}
  >
    <section>
      <label
        htmlFor="txtDescription"
      >
        Description
      </label>
      <input
        autoFocus={true}
        name="txtDescription"
        onChange={[Function]}
        type="text"
        value=""
      />
    </section>
    <section>
      <label
        htmlFor="txtAmount"
      >
        Amount
      </label>
      <input
        name="txtAmount"
        onChange={[Function]}
        type="text"
        value=""
      />
    </section>
    <section>
      <label
        htmlFor="date"
      >
        Date
      </label>
      <SingleDatePicker
        anchorDirection="right"
        customCloseIcon={null}
        customInputIcon={null}
        date={"1970-01-01T00:00:00.000Z"}
        daySize={39}
        disabled={false}
        displayFormat={[Function]}
        enableOutsideDays={false}
        firstDayOfWeek={null}
        focused={false}
        hideKeyboardShortcutsPanel={false}
        horizontalMargin={0}
        id="date"
        initialVisibleMonth={null}
        inputIconPosition="before"
        isDayBlocked={[Function]}
        isDayHighlighted={[Function]}
        isOutsideRange={[Function]}
        isRTL={false}
        keepOpenOnDateSelect={false}
        monthFormat="MMMM YYYY"
        navNext={null}
        navPrev={null}
        numberOfMonths={1}
        onClose={[Function]}
        onDateChange={[Function]}
        onFocusChange={[Function]}
        onNextMonthClick={[Function]}
        onPrevMonthClick={[Function]}
        openDirection="down"
        orientation="horizontal"
        phrases={
          Object {
            "calendarLabel": "Calendar",
            "chooseAvailableDate": [Function],
            "clearDate": "Clear Date",
            "closeDatePicker": "Close",
            "dateIsUnavailable": [Function],
            "enterKey": "Enter key",
            "escape": "Escape key",
            "hideKeyboardShortcutsPanel": "Close the shortcuts panel",
            "homeEnd": "Home and end keys",
            "jumpToNextMonth": "Move forward to switch to the next month",
            "jumpToPrevMonth": "Move backward to switch to the previous month",
            "keyboardNavigationInstructions": "Press the down arrow key to interact with the calendar and
            select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
            "keyboardShortcuts": "Keyboard Shortcuts",
            "leftArrowRightArrow": "Right and left arrow keys",
            "moveFocusByOneDay": "Move backward (left) and forward (right) by one day",
            "moveFocusByOneMonth": "Switch months",
            "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week",
            "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week",
            "openThisPanel": "Open this panel",
            "pageUpPageDown": "page up and page down keys",
            "questionMark": "Question mark",
            "returnFocusToInput": "Return to the date input field",
            "selectFocusedDate": "Select the date in focus",
            "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel",
            "upArrowDownArrow": "up and down arrow keys",
          }
        }
        placeholder="Date"
        readOnly={false}
        renderCalendarInfo={null}
        renderDay={null}
        renderMonth={null}
        reopenPickerOnClearDate={false}
        required={false}
        screenReaderInputMessage=""
        showClearDate={false}
        showDefaultInputIcon={false}
        weekDayFormat="dd"
        withFullScreenPortal={false}
        withPortal={false}
      />
    </section>
    <section>
      <label
        htmlFor="txtNote"
      >
        Note
      </label>
      <input
        name="txtNote"
        onChange={[Function]}
        type="text"
        value=""
      />
    </section>
    <button
      id="btnSubmit"
      type="submit"
    >
      Submit this expense
    </button>
  </form>
</div>
`;
